File: 0079-word-search/0079-word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        def valid(row: int, col: int) -> bool:
            return 0 <= row < m and 0 <= col < n
        
        def backtrack(row: int, col: int, i: int, seen: Set[Tuple[int]]) -> bool:
            if i == len(word):
                return True
            
            for n_r, n_c in (row + 1, col), (row - 1, col), (row, col + 1), (row, col - 1):
                if valid(n_r, n_c) and (n_r, n_c) not in seen:
                    if board[n_r][n_c] == word[i]:
                        seen.add((n_r, n_c))
                        if backtrack(n_r, n_c, i + 1, seen):
                            return True
                        seen.remove((n_r, n_c))
            return False
        
        m, n = len(board), len(board[0])
        for row in range(m):
            for col in range(n):
                if board[row][col] == word[0] and backtrack(row, col, 1, {(row, col)}):
                    return True
        return False

File: 0079-word-search/0079-word-search.cpp
class Solution {
private:
    bool arr[7][7] = {0};
    bool ans(vector<vector<char>>& board, int m, int n, string& s, int i, int j, int pos){
        if (pos == s.size()) return true;
        if (i >= m || i < 0 || j >= n || j < 0 || board[i][j] != s[pos] || arr[i][j] == 1) return false;
        arr[i][j] = 1;
        bool top = ans(board, m, n, s, i - 1, j, pos + 1);
        bool down = ans(board, m, n, s, i + 1, j, pos + 1);
        bool left = ans(board, m, n, s, i, j + 1, pos + 1);
        bool right = ans(board, m, n, s, i, j - 1, pos + 1);
        arr[i][j] = 0;
        return top || down || left || right;
    }
    
public:
    bool exist(vector<vector<char>>& board, string word) {
        int m = board.size(), n = board[0].size();
        int first = 0, last = 0;
        for (int i = 0; i < m; ++i){
            for (int j = 0; j < n; ++j){
                if (board[i][j] == word[0]) ++first;
                else if (board[i][j] == word[word.size() - 1]) ++last; 
            }
        }
        
        if (first > last) reverse(word.begin(), word.end());
        
        for (int i = 0; i < m; ++i){
            for (int j = 0; j < n; ++j){
                if (board[i][j] == word[0]){
                    if (ans(board, m, n, word, i, j, 0)) return true;
                }
            }
        }
        
        return false;
    }
};